### PR TITLE
[Rust] Implement Clone trait for external instances

### PR DESCRIPTION
**Pull request by apepkuss**
In this PR, the `Clone` trait is implemented for `Func`, `Memory`, `Global` and `Table` external instances.

### CODE REVIEW

**Comment by apepkuss** [comment]
@hydai Could you please help review this PR? Thanks a lot!

File: bindings/rust/wasmedge-sys/src/instance/table.rs [hunk 1]
```diff
@@ -153,10 +154,16 @@ impl Table {
...
-            }
+        if !self.registered && Arc::strong_count(&self.inner) == 1 && !self.inner.0.is_null() {
+            unsafe { ffi::WasmEdge_TableInstanceDelete(self.inner.0) };
+        }
+    }
+}
+impl Clone for Table {
+    fn clone(&self) -> Self {
+        Table {
+            inner: self.inner.clone(),
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/instance/table.rs:165]: It looks like the `clone` will only copy the pointer. Is this a expected behavior?

File: bindings/rust/wasmedge-sys/src/instance/memory.rs [hunk 2]
```diff
@@ -248,11 +237,19 @@ impl Memory {
...
-        if !self.registered && !self.inner.0.is_null() {
+        if !self.registered && Arc::strong_count(&self.inner) == 1 && !self.inner.0.is_null() {
             unsafe { ffi::WasmEdge_MemoryInstanceDelete(self.inner.0) };
         }
     }
 }
+impl Clone for Memory {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/instance/memory.rs:248]: I am not familiar with this one(`Arc`). But is it fine to share the pointer(C-level memory pointer) between multiple duplicated memory? Will it cause any data lock issue?

**Review by hydai — CHANGES_REQUESTED** [review]

File: bindings/rust/wasmedge-sys/src/instance/table.rs (hunk 1 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/table.rs:165]: Actually, the `clone` operation does not copy the pointer. The type of `self.inner` is `Arc<InnerTable>`, and `Arc<T>` is a reference type and uses an inner reference counter to record how many references it has at some time. The `clone` method of `Arc<T>` just increments the counter, but does not copy the low-level raw pointer. Thanks!

Comment (hydai) [reply bindings/rust/wasmedge-sys/src/instance/table.rs:165]: Good to know. Thanks.

**Review by hydai — APPROVED** [review]

File: bindings/rust/wasmedge-sys/src/instance/memory.rs (hunk 2 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/memory.rs:248]: There is no data lock issue, as `Arc` is just an atomic reference counter and has no locking mechanism. Thanks a lot!

**Comment by apepkuss** [comment]
@hydai Thanks for the review!

**Pull request merged**